distribution/push_v2.go

```diff
@@ -425,26 +417,17 @@ func (pd *v2PushDescriptor) uploadUsingSession(
 	diffID layer.DiffID,
 	layerUpload distribution.BlobWriter,
 ) (distribution.Descriptor, error) {
-	arch, err := pd.layer.TarStream()
+	reader, release, err := pd.layer.Content(ctx, progressOutput, pd.ID(), schema2.MediaTypeLayer)
```

Comment:
I agree, though this does effect manifest building vs using an existing manifest. I think we can treat that separately and not have it effect this interface. 

This has already got a big bloated on the return value. Might be able to extract this
`Content(ctx context.Context, progressOutput progress.Output, progressKey string, mediaType string) (io.ReadCloser, func(), error)`

If we go that route we can probably just split it out into
```
Open() (io.ReadCloser, error)
Size() int64
MediaType() string
```

Then have the release function either integrated into the closer or register a function to be called on the existing `Release` method.